src/Acuminator/Acuminator.Utilities/Roslyn/Semantic/PXGraph/PXGraphSemanticModel.cs

```diff
@@ -0,0 +1,195 @@
+                }
+            }
+
+            Initializers = initializers.ToImmutableArray();
+        }
+
+        /// <summary>
+        /// Returns one or multiple semantic models of PXGraph and PXGraphExtension descendants which are inferred from <paramref name="typeSymbol"/>
+        /// </summary>
+        /// <param name="pxContext">Context instance</param>
+        /// <param name="typeSymbol">Symbol which is PXGraph or PXGraphExtension descendant and/or which uses PXGraph.InstanceCreated AddHandler method</param>
+        /// <param name="semanticModel">Semantic model</param>
+        /// <param name="cancellation">Cancellation</param>
+        /// <returns></returns>
+        public static IEnumerable<PXGraphSemanticModel> InferModels(PXContext pxContext, INamedTypeSymbol typeSymbol,
+                                                                    CancellationToken cancellation = default)
+        {
+            cancellation.ThrowIfCancellationRequested();
+            pxContext.ThrowOnNull(nameof(pxContext));
+            typeSymbol.ThrowOnNull(nameof(typeSymbol));
+
+            List<PXGraphSemanticModel> models = new List<PXGraphSemanticModel>();
+            GraphType graphType = GraphType.None;
+
+            if (typeSymbol.IsPXGraph(pxContext))
+            {
+                graphType = GraphType.PXGraph;
+            }
+            else if (typeSymbol.IsPXGraphExtension(pxContext))
+            {
+                graphType = GraphType.PXGraphExtension;
+            }
+
+            if (graphType != GraphType.None)
+            {
+                PXGraphSemanticModel explicitModel = new PXGraphSemanticModel(pxContext, graphType, typeSymbol, cancellation);
+
+                models.Add(explicitModel);
+            }
+
+            IEnumerable<InitDelegateInfo> delegates = GetInitDelegates(cancellation, pxContext, typeSymbol);
+
+            foreach (InitDelegateInfo d in delegates)
+            {
+                GraphInitializerInfo info = new GraphInitializerInfo(GraphInitializerType.InstanceCreatedDelegate, d.DelegateNode, d.DelegateSymbol);
+                PXGraphSemanticModel existingModel = models.FirstOrDefault(m => m.Symbol.Equals(d.GraphTypeSymbol));
+                PXGraphSemanticModel implicitModel;
+
+                if (existingModel != null)
+                {
+                    implicitModel = existingModel;
+                }
+                else
+                {
+                    implicitModel = new PXGraphSemanticModel(pxContext, d.GraphType, d.GraphTypeSymbol, cancellation);
+                    models.Add(implicitModel);
+                }
+
+                implicitModel.Initializers = implicitModel.Initializers.Add(info);
+            }
+
+            return models;
+        }
+
+        private static IEnumerable<InitDelegateInfo> GetInitDelegates(CancellationToken cancellation, PXContext pxContext, INamedTypeSymbol typeSymbol)
```

Comment:
@metallmonster @dropsonic should we account for inheritance here? Should the graph semantic model store info only for its type or also for the base types (like SOInvoiceEntry is derived from ARInvoiceEntry). Maybe there should be a flag passed to the factory method?

Reply:
I think that's not needed for initializers. But it makes sense for data views / actions.